### PR TITLE
Add support for TokenFileWebIdentityCredentials

### PR DESCRIPTION
**Pull request by thomaschaaf**
We use Kubernetes to deploy our serverless application for this we need Token File Web Identity Credentials support.

I think the only way to verify this is by setting up a kubernetes cluster and then creating a service role.

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -52,8 +52,8 @@ const impl = {
    * @param results The results to add the given credentials to if they are valid
    * @param credentials The credentials to validate and add to the results if valid
    */
-  addCredentials: (results, credentials) => {
-    if (impl.validCredentials(credentials)) {
+  addCredentials: (results, credentials, skipValidation = false) => {
+    if (skipValidation || impl.validCredentials(credentials)) {
```
[the remark this answers is not in the document]
Comment (thomaschaaf) [reply lib/plugins/aws/provider/awsProvider.js]:
```
> console.log(JSON.stringify(credentials));
{"credentials":{"expired":false,"expireTime":null,"refreshCallbacks":[],"data":null,"clientConfig":{}}}```

The credentials are calculated dynamically with each request by the aws sdk. I think there simply is no way to verify until it tries to make contact with AWS.

**Comment by thomaschaaf** [comment]
@medikoo I have refactored it a little bit and added some documentation.

**Comment by thomaschaaf** [comment]
@medikoo I don't understand what's failing I have not changed anything in the package.json

**Comment by medikoo** [comment]
@thomaschaaf how this token in handled, when operating on bare AWS SDK, as e.g. in below call:

```javascript
const AWS = require("aws-sdk");

const lambda = new AWS.Lambda({ region: "us-east-1" });

lambda.invoke({..}).promise().then(console.log, console.error);
```

Will AWS behind the scenes recognized `AWS_WEB_IDENTITY_TOKEN_FILE` env var and create `AWS.TokenFileWebIdentityCredentials` internally (as e.g. it happens with `AWS_PROFILE`, where it's read transparently behind the scenes and credentials are resolved ).

Or do we need to explicitly create instance of `AWS.TokenFileWebIdentityCredentials` and pass it to SDK constructor, knowing that necessary env vars are there?

**Comment by medikoo** [comment]
> @medikoo I don't understand what's failing I have not changed anything in the package.json

We also didn't change anything. It appears that some npm bug got exposed. I've updated CI setup in master, please update with master and issue will be gone

**Comment by thomaschaaf** [comment]
> @thomaschaaf how this token in handled, when operating on bare AWS SDK, as e.g. in below call:
> 
> ```js
> const AWS = require("aws-sdk");
> 
> const lambda = new AWS.Lambda({ region: "us-east-1" });
> 
> lambda.invoke({..}).promise().then(console.log, console.error);
> ```
> 
> Will AWS behind the scenes recognized `AWS_WEB_IDENTITY_TOKEN_FILE` env var and create `AWS.TokenFileWebIdentityCredentials` internally (as e.g. it happens with `AWS_PROFILE`, where it's read transparently behind the scenes and credentials are resolved ).
> 
> Or do we need to explicitly create instance of `AWS.TokenFileWebIdentityCredentials` and pass it to SDK constructor, knowing that necessary env vars are there?

Nothing needs to be done I am already using this pr happily :) It just magically uses the auth if the env variables are set.

**Comment by medikoo** [comment]
@thomaschaaf sorry I think you misunderstood my question :)

I was asking, whether when working with AWS SDK directly (not through Serverless Framework), you have to create such token and pass it to SDK constructor (or maybe AWS SDK will handle that internally without any action)

**Comment by thomaschaaf** [comment]
> @thomaschaaf sorry I think you misunderstood my question :)
> 
> I was asking, whether when working with AWS SDK directly (not through Serverless Framework), you have to create such token and pass it to SDK constructor (or maybe AWS SDK will handle that internally without any action)

Without my PR I am getting:

```
Serverless Error ---------------------------------------

  AWS provider credentials not found. Learn how to set up AWS provider credentials in our docs here: <http://slss.io/aws-creds-setup>.
```

and if I disable this check: https://github.com/serverless/serverless/blob/f8ee0279037ba35b4c32f5872fcff4e741898db1/lib/plugins/aws/lib/validate.js#L44 the SDK handles it natively.

Alternatively I could write a PR which allows to disable the credential validation.

**Comment by thomaschaaf** [comment]
@medikoo I forgot to mention you above

**Comment by thomaschaaf** [comment]
Now I understand. I could add the check here: https://github.com/serverless/serverless/blob/master/lib/plugins/aws/lib/validate.js#L35

**Comment by medikoo** [comment]
@thomaschaaf I think what you prepare looks good. Still before I agree on that patch, I'd love to know an answer to question I posted above:

> When working with AWS SDK directly (not through Serverless Framework), do you have to create such token and pass it to AWS SDK constructor (or maybe AWS SDK will handle that internally without any action)?

Note: I'm focusing on  case when you want to issue single AWS SDK request (Serverless Framework is not involved), how this authentication is handled then?

Should instance of `AWS.TokenFileWebIdentityCredentials` be created explicitly and passed to SDK constructor, or will AWS behind the scenes recognize `AWS_WEB_IDENTITY_TOKEN_FILE` env var and create `AWS.TokenFileWebIdentityCredentials` internally

**Comment by thomaschaaf** [comment]
Closed for #7442

**Pull request closed without merging**